docs/development/ceps/proposed/cep-002.rst

```diff
@@ -0,0 +1,194 @@
+========
+
+Currently, the hierarchy of the ctapipe ``ArrayEventContainer`` container has the data
+levels first and then each data level has a ``Map`` ``.tel`` for telescope specific
+information in subcontainers.
+This CEP proposes to change this structure to have a container ``TelescopeEventContainer``
+as parent for all telescope-wise information that then contains the data levels.
+
+This has the following advantages:
+
+* At lower data levels, many processing steps can be performed independently for each
+  telescope event, which is easier if all data from a single telescope is in a single container.
+* The current proposed ACADA-DPPS ICD fore-sees event and monitoring files per telescope.
+  This maps nicely to readers filling a ``TelescopeEventContainer`` which are then
+  joined together with array level information into an ``ArrayEventContainer``.
+
+Proposed new structure
+======================
+
+This CEP proposes to change the current layout of ``ArrayEventContainer`` from having multiple
+data levels each with a ``Map`` containing telescope-wise data to a structure
+where each ``ArrayEventContainer`` is composed of one to many ``TelescopeEvents`` containing
+all telescope-wise information for all data levels.
+
+The ``ArrayEventContainer`` should also be renamed to ``SubarrayEventContainer``, to match with other naming
+patterns in ctapipe, such as the ``SubarrayDescription``, making it clear that the array is split
+into multiple subarrays, each observing their own observation block.
+
+The main structure after the change will look like this (Container-suffix left out for readability):
+
+.. code-block::
+
+   SubarrayEvent
+   - index: SubarrayEventIndex
+   - trigger: SubarrayTrigger
+   - simulation: SimulatedShower
+   - dl0: DL0Subarray
+   - dl1: DL1Subarray
+   - dl2: DL2Subarray
+   - dl3: DL3Subarray
+   - tel: Map[int, TelescopeEvent] # or just a list? TelescopeEvent stores the tel_id
+
+   TelescopeEvent
+   - index: TelescopeEventIndex
+   - trigger: TelescopeTrigger
+   - simulation: TelescopeSimulation
+   - r0: R0Telescope
+   - r1: R1Telescope
+   - dl0: DL0Telescope
+   - dl1: DL1Telescope
+   - dl2: DL2Telescope
+   - dl3: DL3Telescope
+
+Which each data level container having specific fields and/or subcontainers including monitoring
+information (interpolated / chosen for that specific event).
+
+
+Advantages of the new structure
+===============================
+
+The new proposed scheme makes it easier to parallelize over array events and move loops
+over telescopes out of code paths that only deal with a single telescope.
+E.g. in the ``CameraCalibrator``, ``ImageProcessor`` and more classes,
+we currently have to provide the ``ArrayEventContainer``,
+so that higher data levels can be filled from lower data levels, although only one telescope
+is processed at a time.
+
+In the current scheme (simplified from ctapipe-process) it looks like this:
+
+.. code-block:: python
+
+   for array_event in source:
```

Comment:
I think it is worth lifting up the hidden loops and writing them out explicitly in this example to make the example more clear, even if it means that the depiction of the event loop is less accurate.

Reply:
done